### DISCUSSION TITLE
Add tablet rotation

### DISCUSSION THREAD

**Discussion by vgskye** [niri-wm/niri#472] [Ideas]
currently tablets(as in digitizers) are, as far as I can see, locked to their "native" rotation. this may be undesirable when, say, it's an on-screen digitizer and the native orientation doesn't match the screen's. (no, map-to-output does not help, it does not rotate it the same way in tablet though it does in touch)

**Comment by YaLTeR** [comment]
What orientation difference are we talking, just 180 degrees, just 90 degrees, are flips needed? There's a `left-handed` property in tablets which may or may not help

**Comment by vgskye** [comment]
for me plain 90deg, but I feel like it should have the options of output transforms for generality

**Reply by YaLTeR** [reply]
Could you send a photo of how it currently looks? Curious

**Reply by vgskye** [reply]
mmh, I'm not sure if I can send a picture. currently with map-to-output unset the digitizer input is stretched to map to the screen bounds but rotated(e.g. pen in top right corner puts cursor in top left, same with bottom-left/bottom-right) and with map-to-output set the digitizer's aspect ratio is seemingly preserved with the cursor hitting the bottom edge of the screen when the pen is in the left half.

**Reply by YaLTeR** [reply]
What if you try to set transform on the output itself? https://github.com/YaLTeR/niri/wiki/Configuration:-Outputs#transform

**Reply by vgskye** [reply]
ah, yeah, that's the issue, the output is transformed 90deg using that but the tablet input is not.

**Reply by vgskye** [reply]
if I don't set a transform on output, the tablet input works fine but the entire screen is, well, sideways

**Reply by YaLTeR** [reply]
I see. How does it work in GNOME?

**Reply by vgskye** [reply]
...hell if I know, I've not run GNOME in this device in months

**Comment by vgskye** [comment]
for future reference, I was able to hack around this by configuring libinput via udev hwdb, but a native support would be far better

**Reply by dchang0** [reply]
I agree--native support in Niri would be far better, especially for different monitors. For instance, I have a dual-monitor setup where one monitor can turn 90 degrees for portrait layout. And I also have a LINUX tablet where having the accelerometer rotate the screen (and this should be lockable to a particular orientation) would be extremely useful.

I used to use Ubuntu 24 with GNOME but got tired of stacking windows. That said, its native support for tablet-accelerometer rotation was very good.

**Reply by Sempyos** [reply]
I've been using this for a while and found it and adequate solution: https://github.com/Lassebq/wlr-autorotate

**Reply by dchang0** [reply]
Thanks, I will check that out!

**Reply by KyleKrein** [reply]
> I've been using this for a while and found it and adequate solution: https://github.com/Lassebq/wlr-autorotate

For me wlr-autorotate only rotates the compositor itself, but not the tablet (digitizer). So my stylus doesn't work correctly in any rotation other than "normal". I'd really appreciate a solution, that will work for the digitizer...

**Reply by dchang0** [reply]
> > I've been using this for a while and found it and adequate solution: https://github.com/Lassebq/wlr-autorotate
> 
> For me wlr-autorotate only rotates the compositor itself, but not the tablet (digitizer). So my stylus doesn't work correctly in any rotation other than "normal". I'd really appreciate a solution, that will work for the digitizer...

Ohhhh, that would mean I can't use wlr-autorotate yet either... I wonder how Ubuntu 24+GNOME handled it... Maybe I'll re-install those on the tablet and figure out which piece of software they used.

**Reply by HiImMuffin** [reply]
> for future reference, I was able to hack around this by configuring libinput via udev hwdb, but a native support would be far better

May I ask how did you configure it?

**Comment by Xananax** [comment]
Same problem, I can rotate, but the stylus won't work. I got this in my binds:

```kdl
    Ctrl+Alt+Left { spawn "niri" "msg" "output" "eDP-1" "transform" "90"; }
    Ctrl+Alt+Right { spawn "niri" "msg" "output" "eDP-1" "transform" "270"; }
    Ctrl+Alt+Down { spawn "niri" "msg" "output" "eDP-1" "transform" "normal"; }
    Ctrl+Alt+Up { spawn "niri" "msg" "output" "eDP-1" "transform" "180"; }
```

<details>
<summary>Here is me drawing on Krita.  (open for image)</summary>
<img width="1801" height="2743" alt="image" src="https://github.com/user-attachments/assets/ad3e4ee7-5b94-4a69-b198-e95651594254" />
In yellow: what I drew with my pen on the screen. In blue: what was actually drawn.
</details>

The shortcuts work well, but then the pen remains in the original orientation, making drawing impossible. I wonder if there's a solution or workaround for now.

**Comment by oisanjaya** [comment]
Check out #3763 i'm trying to implement auto rotation into niri
